content/blog/2022/2022-04-26-introduction-to-retrofit.md

```diff
@@ -0,0 +1,536 @@
+````java
+    @PUT("/library/managed/books/{id}")
+    Call<LibResponse> updateBook(@Path("id") Long id, @Body BookDto book);
+````
+To pass the actual value of `id`, we set a method parameter `@Path` so that the call execution will replace `{id}` with its corresponding value.
+
+We can specify the query parameters in the URL directly or add `@Query` param to the method.
+````java
+    @GET("/library/managed/books?type=all")
+    OR
+    @GET("/library/managed/books")
+    Call<List<BookDto>> getAllBooks(@Query("type") String type);
+````
+If the request needs to have multiple query parameters, we could use `@QueryMap`
+````java
+   @GET("/library/managed/books")
+    Call<List<BookDto>> getAllBooks(@QueryMap Map<String, String> options);
+````
+To specify an object as HTTP request body, we use the `@Body` annotation.
+````java
+    @POST("/library/managed/books")
+    Call<LibResponse> createNewBook(@Body BookDto book);
+````
+To the Retrofit interface methods, we can specify static or dynamic header parameters
+For static headers, we have `@Headers`
+````java
+    @Headers("Accept: application/json")
+    @GET("/library/managed/books")
+    Call<List<BookDto>> getAllBooks(@Query("type") String type);
+````
+We could also define multiple static headers as:
+````java
+    @Headers({
+        "Accept: application/json",
+        "Cache-Control: max-age=640000"})
+    @GET("/library/managed/books")
+    Call<List<BookDto>> getAllBooks(@Query("type") String type);
+````
+To pass dynamic headers, we specify them as method parameters
+````java
+    @GET("/library/managed/books/{requestId}")
+    Call<BookDto> getAllBooksWithHeaders(@Header("requestId") String requestId);
+````
+For multiple dynamic headers, we use `@HeaderMap`.
+
+**All Retrofit responses are wrapped in a `Call` object. It supports both blocking and non-blocking requests.**
+
+### Using the Retrofit Builder API
+The Builder API on Retrofit allows for customization of the Configuration object. We will take a closer look at some configuration options.
+
+#### Configuring timeout settings
+We can set timeouts on the underlying Http client. However, setting up these values is optional. If we do not specify the timeouts, default settings apply.
+- Connection timeout: 10 sec
+- Read timeout: 10 sec
+- Write timeout: 10 sec
+To override these defaults, we need to setup `OkHttpClient` as shown below:
+````java
+        OkHttpClient.Builder httpClientBuilder = new OkHttpClient.Builder()
+                .connectTimeout(props.getConnectionTimeout(), TimeUnit.SECONDS)
+                .readTimeout(props.getReadWriteTimeout(), TimeUnit.SECONDS);
+
+        return new Retrofit.Builder().client(httpClientBuilder.build())
+                .baseUrl(props.getEndpoint())
+                .addConverterFactory(JacksonConverterFactory.create(new ObjectMapper()))
+                .build().create(LibraryClient.class);
+````
+Here, the timeout values are as specified in application.yaml.
```

Comment:
Like I mentioned above, please provide a code example of the relevant part of the application.yml (in the code example of my previous comment about this, maybe just point to this place here)